### PR TITLE
Add native LFM2-Audio support

### LINKED ISSUE #43909: Add LFM2.5 Audio 1.5B
### Model description

I would like to add LFM2.5 Audio, which is a highly versatile multimodal audio-text model for its small size of 1.5b. Just ran the model on cpu and it's really good. Should be a good candidate for transformers integration. @eustlb @ebezzam 

The modular implementation should not be very long since the original repo directly uses lfm2 text model from transformers and they use mimi from the original moshi repo (should be possible to use the transformer native module). They use fastconformer for the encoder (slightly modified scripts from Nemo), but I'm not very sure if there's a module for that in transformers (parakeet might have it). 

### Open source status

- [x] The model implementation is available
- [x] The model weights are available

### Provide useful links for the implementation

authors: @haerski, @ArthurBook
[original repo](https://github.com/Liquid4All/liquid-audio)
[weights](https://huggingface.co/LiquidAI/LFM2.5-Audio-1.5B)

If permitted, I'll go ahead and open a PR.

### PR DESCRIPTION
**Pull request by kadirnar**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48249) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48249)

## What does this PR do?

Adds native Transformers support for `LiquidAI/LFM2.5-Audio-1.5B`:

- ASR, TTS, and sequential/interleaved text-audio generation
- native `ParakeetEncoderModel` reuse for FastConformer
- bundled LFM audio detokenizer, with native `MimiModel` fallback for older checkpoints
- no copied Moshi source directory

Fixes #43909.

## Reference parity

Compared against `Liquid4All/liquid-audio` v1.3.0 (`19e6584`) with the same checkpoint, prompts, audio inputs, CUDA bf16 execution, and greedy decoding (`top_k=1`).

| Task | Exact result |
| --- | --- |
| ASR | 58 / 58 generation events |
| TTS — US male | 112 / 112 events; 110 / 110 audio frames |
| TTS — US female | 512 / 512 events; 511 / 511 audio frames |
| TTS — UK male | 512 / 512 events; 511 / 511 audio frames |
| TTS — UK female | 106 / 106 events; 104 / 104 audio frames |
| Interleaved chat — turn 1 | 512 / 512 events; 54 text tokens and 458 audio frames |
| Interleaved chat — turn 2 | 378 / 378 events; 59 text tokens and 319 audio frames |
| Audio detokenizer | Exact waveform tensor; max absolute difference `0` |

All eight audio codebooks matched exactly. The main checkpoint and bundled detokenizer both load with zero missing, unexpected, or mismatched keys.

<details>
<summary>🤖 mlinter review state</summary>

- hash: `3d87dcff`
- commit: `e525641c`
- review: https://github.com/huggingface/transformers/pull/48249#pullrequestreview-5254929565

</details>

### CODE REVIEW

**Comment by kadirnar** [comment]
@eustlb can you check it? I tested it locally and it works.

File: docs/source/en/model_doc/lfm2_audio.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+from transformers import Lfm2AudioForConditionalGeneration, Lfm2AudioProcessor
+
+
+model_id = "LiquidAI/LFM2.5-Audio-1.5B"
+processor = Lfm2AudioProcessor.from_pretrained(model_id)
+model = Lfm2AudioForConditionalGeneration.from_pretrained(
+    model_id,
+    device_map="auto",
+    dtype=torch.bfloat16,
+)
```
Comment (eustlb) [inline docs/source/en/model_doc/lfm2_audio.md]:
```suggestion
    device_map="auto",
)
```
"dtype": "bfloat16" should be in config

**Comment by ebezzam** [comment]
Thanks @kadirnar for the PR! I'll try to get back to you next week with a first set of comments

File: docs/source/en/model_doc/lfm2_audio.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+1. a log-mel frontend and a 17-layer FastConformer audio encoder,
+2. a small adapter that projects encoded speech to the LFM2 hidden size,
+3. an LFM2 backbone that processes text, input-audio, and output-audio positions in one timeline, and
+4. a six-layer depth transformer that predicts eight Mimi codebook tokens for each generated audio frame.
+
+The implementation reuses the native Transformers [`ParakeetEncoderModel`] for the FastConformer. It does not depend
+on Moshi or duplicate its source code in an LFM2-Audio-specific folder. For waveform decoding,
+`LiquidAI/LFM2.5-Audio-1.5B` includes a compact LFM detokenizer that is loaded lazily by
+[`Lfm2AudioProcessor.decode_audio`]. A native Transformers [`MimiModel`] remains available as a fallback for older
+checkpoints.
```
Comment (ebezzam) [inline docs/source/en/model_doc/lfm2_audio.md]: nit: these details can be removed
```suggestion
```

File: src/transformers/models/lfm2_audio/configuration_lfm2_audio.py [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+
+from ...configuration_utils import PreTrainedConfig
+from ...utils import auto_docstring
+from ..lfm2.configuration_lfm2 import Lfm2Config
+from ..parakeet.configuration_parakeet import ParakeetEncoderConfig
+
+
+@auto_docstring(checkpoint="LiquidAI/LFM2.5-Audio-1.5B")
+@strict
+class Lfm2AudioPreprocessorConfig(PreTrainedConfig):
```
Comment (ebezzam) [inline src/transformers/models/lfm2_audio/configuration_lfm2_audio.py]: anything on preparing inputs (mel computation and padding to batch) should be part of the feature extractor, and not in the modeling. For example, here is Parakeet's feature extractor: https://github.com/huggingface/transformers/blob/main/src/transformers/models/parakeet/feature_extraction_parakeet.py

And a model can use another model's feature extractor (for example in the mapping [here](https://github.com/huggingface/transformers/blob/719d8090f4040de8563ccf05496419b212fb395f/src/transformers/models/auto/feature_extraction_auto.py#L40)), which may be the case here if you're using something similar to Parakeet's encoder

File: docs/source/en/model_doc/lfm2_audio.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+## Usage
+
+The original `LiquidAI/LFM2.5-Audio-1.5B` checkpoint predates its native Transformers integration. Use the concrete
+classes below to load that checkpoint directly.
```
Comment (ebezzam) [inline docs/source/en/model_doc/lfm2_audio.md]: it would be great to use the original checkpoint, but we'll likely have to create a new one due to changes in the chat template. Thus requiring a conversion+upload script like [this](https://github.com/huggingface/transformers/blob/main/src/transformers/models/vibevoice_asr/convert_vibevoice_asr_to_hf.py), where we initialize the new chat template when creating the processor like [this](https://github.com/huggingface/transformers/blob/main/src/transformers/models/vibevoice_asr/convert_vibevoice_asr_to_hf.py#L251)

File: src/transformers/models/lfm2_audio/processing_lfm2_audio.py [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+
+class Lfm2AudioProcessorKwargs(ProcessingKwargs, total=False):
+    audio_kwargs: Lfm2AudioAudioKwargs
+    _defaults = {}
+
+
+Lfm2AudioProcessorKwargs.__annotations__["audio_kwargs"] = Lfm2AudioAudioKwargs
+
+
+DEFAULT_CHAT_TEMPLATE = r"""{{- bos_token -}}
```
Comment (ebezzam) [inline src/transformers/models/lfm2_audio/processing_lfm2_audio.py]: as mentioned [here](https://github.com/huggingface/transformers/pull/48249/changes#r4047783891), I expected we'll have to change the chat template. So let's do a conversion script and move this out of here.

File: src/transformers/models/lfm2_audio/processing_lfm2_audio.py [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+        processor = super().from_pretrained(pretrained_model_name_or_path, **kwargs)
+        if decoder_model_id is not None:
+            processor.decoder_model_id = str(decoder_model_id)
+        elif processor.decoder_model_id is None:
+            processor.decoder_model_id = str(pretrained_model_name_or_path)
+        return processor
+
+    @classmethod
+    def _get_arguments_from_pretrained(cls, pretrained_model_name_or_path, processor_dict=None, **kwargs):
+        """Build the frontend from `config.json` for checkpoints released before processor files existed."""
```
Comment (ebezzam) [inline src/transformers/models/lfm2_audio/processing_lfm2_audio.py]: we shouldn't overwrite these methods. If we configure the processor in the conversion script like [this](https://github.com/huggingface/transformers/blob/719d8090f4040de8563ccf05496419b212fb395f/src/transformers/models/vibevoice_asr/convert_vibevoice_asr_to_hf.py#L247), it will create a processor like [this](https://huggingface.co/microsoft/VibeVoice-ASR-HF/blob/main/processor_config.json) for proper loading. Please see other models (Canary, Parakeet, Qwen3ASR, VibeVoice ASR,  AudioFlamingo3, etc) for conventions in the `processing_*.py` file

File: src/transformers/models/lfm2_audio/processing_lfm2_audio.py [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+        feature_lengths = feature_attention_mask.sum(-1)
+        # Liquid Audio's inference path uses the complete centered-STFT output, including the terminal frame that
+        # the NeMo frontend zeroes beyond its reported length. That extra frame matters whenever the reported length
+        # is divisible by FastConformer's 8x subsampling factor: omitting it removes one audio placeholder and can
+        # make ASR generation collapse to empty text.
+        feature_lengths = (feature_lengths + 1).clamp_max(feature_attention_mask.shape[-1])
+        frame_indices = torch.arange(feature_attention_mask.shape[-1], device=feature_attention_mask.device)
+        feature_attention_mask = frame_indices[None] < feature_lengths[:, None]
+        audio_inputs["input_features_attention_mask"] = feature_attention_mask
+        audio_inputs["num_audio_tokens"] = (feature_lengths + 7) // 8
```
Comment (ebezzam) [inline src/transformers/models/lfm2_audio/processing_lfm2_audio.py]: if we have to make such changes, let's instead define a new feature extractor. Modular can also be used to generated it in case you only need to change the forward method

File: src/transformers/models/parakeet/feature_extraction_parakeet.py
Comment (ebezzam) [inline src/transformers/models/parakeet/feature_extraction_parakeet.py:1]: let's instead create a new feature extractor for LFM. 

Normally, improvements to Parakeet's would be better as it's own PR, but as there's an [ongoing one](https://github.com/huggingface/transformers/pull/44394) to refactor the feature extractors, let's rather create a new one for LFM.

File: src/transformers/conversion_mapping.py
Comment (ebezzam) [inline src/transformers/conversion_mapping.py:1]: these changes would no longer be needed as we use a conversion script

File: tests/models/parakeet/test_feature_extraction_parakeet.py
Comment (ebezzam) [inline tests/models/parakeet/test_feature_extraction_parakeet.py:1]: these would be new tests for your LFM feature extractor instead

File: tests/models/lfm2_audio/test_modeling_lfm2_audio.py
Comment (ebezzam) [inline tests/models/lfm2_audio/test_modeling_lfm2_audio.py:1]: very important! we are missing integration tests like [this](https://github.com/huggingface/transformers/blob/7301fbc6b082b92bcbe59224a8ae7555dde4fdb7/tests/models/parakeet/test_modeling_parakeet.py#L746) with a reproducer [gist](https://github.com/huggingface/transformers/blob/7301fbc6b082b92bcbe59224a8ae7555dde4fdb7/tests/models/parakeet/test_modeling_parakeet.py#L748). The goal of these files is to compare with usage with the original checkpoint + code, which is how the reproducer computes the expected outputs.

Protip: This is the first thing I do when adding a new model so that as we iteratively clean/prune the Transformers version, we can check that we don't have regression with the original code. E.g. by running the tests like so (for parakeet) when major changes are made:
```
CUDA_VISIBLE_DEVICES=0,1 RUN_SLOW=1 pytest tests/models/parakeet/test_modeling_parakeet.py::ParakeetForTDTIntegrationTest
```

File: src/transformers/models/lfm2_audio/modular_lfm2_audio.py [hunk 8]
```diff
@@ -0,0 +1,977 @@
...
+
+
+TEXT_MODALITY = 1
+AUDIO_INPUT_MODALITY = 2
+AUDIO_OUTPUT_MODALITY = 3
+
+
+@auto_docstring(custom_intro="Base class for LFM2-Audio backbone outputs.")
+@dataclass
+class Lfm2AudioModelOutputWithPast(ModelOutput):
```
Comment (ebezzam) [inline src/transformers/models/lfm2_audio/modular_lfm2_audio.py]: is there a model output from another model we can use via modular?

File: src/transformers/models/lfm2_audio/modular_lfm2_audio.py [hunk 9]
```diff
@@ -0,0 +1,977 @@
...
+    last_hidden_state: torch.FloatTensor | None = None
+    past_key_values: Cache | None = None
+    hidden_states: tuple[torch.FloatTensor, ...] | None = None
+    attentions: tuple[torch.FloatTensor, ...] | None = None
+    audio_hidden_states: torch.FloatTensor | None = None
+
+
+@auto_docstring(custom_intro="Base class for LFM2-Audio conditional-generation outputs.")
+@dataclass
+class Lfm2AudioConditionalGenerationOutput(ModelOutput):
```
Comment (ebezzam) [inline src/transformers/models/lfm2_audio/modular_lfm2_audio.py]: similar as above, can we use modular from another model?

File: src/transformers/models/lfm2_audio/modular_lfm2_audio.py [hunk 10]
```diff
@@ -0,0 +1,977 @@
...
+    key_dtype = key_states.dtype
+    query_complex = torch.view_as_complex(query_states.float().reshape(*query_states.shape[:-1], -1, 2))
+    key_complex = torch.view_as_complex(key_states.float().reshape(*key_states.shape[:-1], -1, 2))
+    frequencies = frequencies.view(1, frequencies.shape[0], 1, frequencies.shape[1])
+    query_states = torch.view_as_real(query_complex * frequencies).flatten(-2)
+    key_states = torch.view_as_real(key_complex * frequencies).flatten(-2)
+    return query_states.to(query_dtype), key_states.to(key_dtype)
+
+
+class Lfm2AudioDepthAttentionCore(nn.Module):
```
Comment (ebezzam) [inline src/transformers/models/lfm2_audio/modular_lfm2_audio.py]: likewise using modular, as attention is already implemented in the lib

**Review by ebezzam — COMMENTED** [review]
@kadirnar thanks for the PR! Here are a first set of comments to iterate on. 

Here are the main points:
1. Create a conversion+upload script to create a Transformers-native version. Due to changes in the chat template, we need a new checkpoint. You can create this checkpoint under your account for now, e.g. `kadirnar/LFM2.5-Audio-1.5B-hf`. And when we are ready to merge we will ask Liquid AI to create a checkpoint under their org (e.g. as we've done with Qwen3 ASR [here](https://huggingface.co/Qwen/Qwen3-ASR-1.7B-hf)).
2. Create integration tests between (1) the new checkpoint `kadirnar/LFM2.5-Audio-1.5B-hf` + Transformers code and (2) expected outputs with the original checkpoint+model. This will be very helpful while we iteratively clean/prune the native Transformers implementation to prevent regression!
3. Update the modular file to "inherit" modules from other models so we can avoid code repetition. Here is some docs on using modular: https://huggingface.co/docs/transformers/main/en/modular_transformers

Finally, we have this [SKILL file](https://huggingface.co/datasets/MyTransformersSkills/self-review-skill/blob/main/transformers_contribution_skill.md) which can help to do a self-review independently. It can help a lot to identify blockers for a Transformers integration.

Hope that helps, and thanks again for your contribution 🤗

**Comment by ebezzam** [comment]
@kadirnar could you also open a submission of the original model to the [Open ASR Leaderboard](https://github.com/huggingface/open_asr_leaderboard/blob/main/.github/PULL_REQUEST_TEMPLATE.md#new-model-checklist)? I'm curious how it compares to other ASR models in terms of performance

**Comment by kadirnar** [comment]
> @kadirnar thanks for the PR! Here are a first set of comments to iterate on.
> 
> Here are the main points:
> 
>     1. Create a conversion+upload script to create a Transformers-native version. Due to changes in the chat template, we need a new checkpoint. You can create this checkpoint under your account for now, e.g. `kadirnar/LFM2.5-Audio-1.5B-hf`. And when we are ready to merge we will ask Liquid AI to create a checkpoint under their org (e.g. as we've done with Qwen3 ASR [here](https://huggingface.co/Qwen/Qwen3-ASR-1.7B-hf)).
> 
>     2. Create integration tests between (1) the new checkpoint `kadirnar/LFM2.5-Audio-1.5B-hf` + Transformers code and (2) expected outputs with the original checkpoint+model. This will be very helpful while we iteratively clean/prune the native Transformers implementation to prevent regression!
> 
>     3. Update the modular file to "inherit" modules from other models so we can avoid code repetition. Here is some docs on using modular: https://huggingface.co/docs/transformers/main/en/modular_transformers
> 
> 
> Finally, we have this [SKILL file](https://huggingface.co/datasets/MyTransformersSkills/self-review-skill/blob/main/transformers_contribution_skill.md) which can help to do a self-review independently. It can help a lot to identify blockers for a Transformers integration.
> 
> Hope that helps, and thanks again for your contribution 🤗

Thank you for the detailed explanation and help. I have made the corrections.
